### ISSUE TITLE
Replace rollup-stream with raw rollup

### ISSUE THREAD

**Issue by mistressofjellyfish** [mistressofjellyfish/not-grocy#2]
Since viewjs are now functions, bundling becomes *much* easier and with less configuration, so remove gulp and just use webpack directly.

**Comment by mohe2015** [comment]
Maybe https://vitejs.dev/ should also get a chance as it's been originally written for Vue and may work pretty well. But this would be another new tool to learn for you (probably).

**Comment by mistressofjellyfish** [comment]
eh, I'm a fast learner. vite looks incredibly cool, but I currently don't see how it integrates with the legacy php views.

Once they're gone though, this will be a godsend. But I will not start a new frontend from scratch, but progressively adapt vue. I've had enough bad "this needs to be rewritten from scratch" experiences in the past that I will prefer a somewhat incremental route when I can.

But I think opening an issue and putting it in a project "future" is worth the effort, just so I won't forget it.

**Comment by mohe2015** [comment]
Crossref https://github.com/mistressofjellyfish/not-grocy/issues/9

**Comment by mistressofjellyfish** [comment]
After having sunk another day into trying to get it done better with webpack and getting absolutely nowhere, I'll think I'll leave it at this for the moment.

I had better success when I toyed around with using raw rollup though, so let's see if something comes from that, if not, I'll leave gulp in for the moment.

**Comment by mistressofjellyfish** [comment]
https://gulpjs.com/docs/en/getting-started/using-plugins#do-you-need-a-plugin See also.

However, last time I checked I had some issues with css extraction and putting the extracted files into the CSS directory (which required a separate moving step – not the end of the world, but not particularly pretty). But honestly, that isn't really a problem anymore. If all gets bundled into one big chunk anyways, might as well do the same for the FullCalendar css files (which are the last remaining conditionally included files).

**Comment by mistressofjellyfish** [comment]
So, currently, using rollup looks actually very promising.

However, now the point has come where I absolutely don't see why we should need gulp. Problem is; I really don't like npm-scripts; `package.json` really sucks for doing semi-complex tasks, and I'm also not a fan of pulling a load of javascript dev dependencies to do stuff like `cp`, `mv` or `tar -cvfj`.

Also I still have an issue with parallelism there. So... What do you think of using 1970 tech to solve 2021 problems and use a bog-standard well-known `Makefile`?

This will break building on Windows shell, but honestly either run the commands directly or just use WSL. I don't want to spin up node for copying files when a simple, well tested C program that had a stable CLI for decades can do that as well.

I also looked at eslint building, but that's probably best saved for the time #9 is looked at.

**Comment by mohe2015** [comment]
> Also I still have an issue with parallelism there. So... What do you think of using 1970 tech to solve 2021 problems and use a bog-standard well-known `Makefile`?

I don't care and if I understand you correctly we would switch to vite or so later anyways. Because using the Makefile for forever seems a little strange/uncommon to me.

**Comment by mistressofjellyfish** [comment]
There are certain tasks that at the moment I don't see how vite solves efficiently. These include (in no particular order, and this list isn't exhaustive):

- language file extraction and generation. AFAIK there is no path from the gettext po files to vue-i18n; so we need a converter (that does at least gettext-po --> vue-i18n, but most probably vue-i18n --> gettext-po as well). For extraction, https://github.com/pixari/vue-i18n-extract needs to be run before building, and that can be done in parallel to any css bundling or asset minification.
- everything that deals with php linting & style checking / formatting
- release tar generation

> I don't care and if I understand you correctly we would switch to vite or so later anyways. Because using the Makefile for forever seems a little strange/uncommon to me.

My career background is in security and everything that deals with native code, so makefiles are quite natural to me. I have no idea if make is common in modern web development – probably not. To me, the idea to spin up node for literally everything feels very strange, especially considering the huge performance impact spinning up node has. In any case, let's re-evaluate the situation when we get to vite.

**Comment by mistressofjellyfish** [comment]
Okay, so I got that working :) Not quite in a state ready to push, but I now have a 200-odd line long makefile that handles *everything*, including conditionally running yarn and composer.

`make clean && make -j8 build` (which *nukes* node_modules/ and vendor/) runs in just over 2 Minutes, a subsequent sass / rollup / postcss build with `make -j8 build` is done in about 24 seconds, where most time is spent compiling & bundling the javascript and all its dependencies. Additional tasks most probably won't come with much of an overhead though as even folder generation is now parallel.

Speed could probably be improved by chunking away all vendor stuff (which is now in grocy.js), but I haven't found a nice solution for this. The main problem for this is how jQuery and its plugins pollute everything and just refuse to work properly in the global space. I already needed to nuke jquery.timeago and replace it with a non-jquery timeago plugin because resolution just didn't work. AFAICT, the rest of the jquery-dependant clusterf- does find the instance. Still have to sort out a localization bug with the non-jquery timeago, but so far, it works. Whatever solution there might be, it only really saves time if it allows us to run rollup in parallel and only conditionally rebuilds when `js/vendor.js` or `js/helpers/global.js` have changed. But because I have migrated quite some stuff away from global to local imports, I'm afraid the "vendor-chunk-will-change"-detection needs to be within rollup and I didn't see how to configure that. In all reality though I can live with 24s build time for now.

Additionally, eslint now brutally enforces its rules, and that includes the gist of the legacy formatting: Tabs for indentation, Allman-Style braces, use semicolons everywhere. This is once again the reason for an absolute monster of a diff. But at least stuff is now a little more consistent, and should be in the future.

For the future, I think we should discuss the code style to use (for both php and javascript), migrate and enforce it. However, I haven't really looked at the common styles out there, so I don't have an opinion on which style to use right now. Whichever we settle on, I don't want to invent my own, if I can get comfortable with an already published style.

`make watch` now also conveniently starts the php backend.
